src/ray/util/url.cc

```diff
@@ -16,64 +16,114 @@
 
 #include <stdlib.h>
 
+#ifdef _WIN32
+#include <afunix.h>
+#else
+#include <sys/un.h>
+#endif
+
 #include <algorithm>
-#include <boost/asio/ip/address.hpp>
-#include <boost/asio/ip/tcp.hpp>
+#include <sstream>
 #include <string>
 
+#include <boost/asio/generic/stream_protocol.hpp>
+#ifndef _WIN32
+#include <boost/asio/local/stream_protocol.hpp>
+#endif
+#include <boost/asio/ip/tcp.hpp>
+
+#include "ray/util/filesystem.h"
 #include "ray/util/logging.h"
+#include "ray/util/util.h"
 
-boost::asio::ip::tcp::endpoint parse_ip_tcp_endpoint(const std::string &endpoint,
-                                                     int default_port) {
-  const std::string scheme_sep = "://";
-  size_t scheme_begin = 0, scheme_end = endpoint.find(scheme_sep, scheme_begin);
-  size_t host_begin;
-  if (scheme_end < endpoint.size()) {
-    host_begin = scheme_end + scheme_sep.size();
-  } else {
-    scheme_end = scheme_begin;
-    host_begin = scheme_end;
+namespace ray {
+
+/// Uses sscanf() to read a token matching from the string, advancing the iterator.
+/// \param c_str A string iterator that is dereferenceable. (i.e.: c_str < string::end())
+/// \param format The pattern. It must not produce any output. (e.g., use %*d, not %d.)
+/// \return The scanned prefix of the string, if any.
+static std::string ScanToken(std::string::const_iterator &c_str, std::string format) {
+  int i = 0;
+  std::string result;
+  format += "%n";
+  if (static_cast<size_t>(sscanf(&*c_str, format.c_str(), &i)) <= 1) {
+    result.insert(result.end(), c_str, c_str + i);
+    c_str += i;
   }
-  std::string scheme = endpoint.substr(scheme_begin, scheme_end - scheme_begin);
-  RAY_CHECK(scheme_end == host_begin || scheme == "tcp");
-  size_t port_end = endpoint.find('/', host_begin);
-  if (port_end >= endpoint.size()) {
-    port_end = endpoint.size();
+  return result;
+}
+
+std::string endpoint_to_url(
+    const boost::asio::generic::basic_endpoint<boost::asio::generic::stream_protocol> &ep,
+    bool include_scheme) {
+  std::string result;
+  switch (ep.protocol().family()) {
+  case AF_INET:
+  case AF_INET6: {
+    boost::asio::ip::tcp protocol = ep.protocol().family() == AF_INET
+                                        ? boost::asio::ip::tcp::v4()
+                                        : boost::asio::ip::tcp::v6();
+    boost::asio::ip::tcp::endpoint e(protocol, 0);
+    RAY_DCHECK(e.size() == ep.size());
+    memcpy(e.data(), ep.data(), ep.size());
```

Comment:
It seems to me that this line makes some assumption about the representation of `boost::asio::ip::tcp::endpoint` and `boost::asio::generic::basic_endpoint`, seems unsafe. There is an explicit way to convert from the tcp endpoint to the basic endpoint in the boost documentation, but I don't think the other way around is supported. Is that right?

Reply:
That's correct regarding the conversion. However, this is just the C++ version of the mandatory cast from `struct sockaddr *` to `struct sockddr_in *` (or `struct sockaddr_un *`) in C, which is the way you're supposed to do it. So I guess it's "implicit" in some sense, but not in the sense that I'd expect it to randomly break with a Boost upgrade.

Reply:
After digging a bit more into this, it seems that you can use this assignment operator to do it in a nicer way:
```
https://github.com/chriskohlhoff/asio/blob/8087252a0c3c2f0baad96ddbd6554db17a846376/asio/include/asio/generic/basic_endpoint.hpp#L98
```

Also,  instead of using the stringstreams below, we can do `result = e.address()`, right?

Reply:
I think that's only assigning an instance of the same endpoint class, not a different one?

Reply:
When you're inside `template <typename Protocol> class basic_endpoint`, `basic_endpoint` by itself just refers to `basic_endpoint<Protocol>`. You'd want `basic_endpoint<Other>` or something like that here for this to work.

Reply:
Then maybe this: https://github.com/chriskohlhoff/asio/blob/8087252a0c3c2f0baad96ddbd6554db17a846376/asio/include/asio/generic/basic_endpoint.hpp#L70? There must be a way to do this without mem copy...

Reply:
Oh interesting, that's not inside the documentation. Should I use it anyway? It might break later...

Reply:
Actually I don't think that's working? In fact now I remember I ran into this compile error when I tried this myself:

	src/ray/util/url.cc(63,36): error: no matching constructor for initialization of 'boost::asio::ip::tcp::endpoint' (aka 'basic_endpoint<boost::asio::ip::tcp>')
	    boost::asio::ip::tcp::endpoint e(ep);
	                                   ^ ~~
	external/boost\boost/asio/ip/basic_endpoint.hpp(97,3): note: candidate constructor not viable: no known conversion from 'const boost::asio::generic::basic_endpoint<boost::asio::generic::stream_protocol>' to 'const boost::asio::ip::basic_endpoint<boost::asio::ip::tcp>' for 1st argument
	  basic_endpoint(const basic_endpoint& other) BOOST_ASIO_NOEXCEPT
	  ^
	external/boost\boost/asio/ip/basic_endpoint.hpp(104,3): note: candidate constructor not viable: no known conversion from 'const boost::asio::generic::basic_endpoint<boost::asio::generic::stream_protocol>' to 'boost::asio::ip::basic_endpoint<boost::asio::ip::tcp>' for 1st argument
	  basic_endpoint(basic_endpoint&& other) BOOST_ASIO_NOEXCEPT
	  ^
	external/boost\boost/asio/ip/basic_endpoint.hpp(60,3): note: candidate constructor not viable: requires 0 arguments, but 1 was provided
	  basic_endpoint() BOOST_ASIO_NOEXCEPT
	  ^
	external/boost\boost/asio/ip/basic_endpoint.hpp(81,3): note: candidate constructor not viable: requires 2 arguments, but 1 was provided
	  basic_endpoint(const InternetProtocol& internet_protocol,
	  ^
	external/boost\boost/asio/ip/basic_endpoint.hpp(90,3): note: candidate constructor not viable: requires 2 arguments, but 1 was provided
	  basic_endpoint(const boost::asio::ip::address& addr,
	  ^